examples/mugen/generation/text_video_gpt.py

```diff
@@ -0,0 +1,225 @@
+    return tuple([s // d for s, d in zip(input_shape, downsample)])
+
+
+class TextTokenizer(nn.Module):
+    """Converts between text and tokens / embedings
+
+    Wrapper around the tokenizer to be consistent with the API required by
+    :py:class:`torchmultimodal.models.gpt.MultimodalGPT`. It also contains the
+    embedding layer to enable lookup by token ids.
+    """
+
+    def __init__(
+        self,
+        context_len: int,
+        d_model: int,
+        tokenizer: nn.Module,
+    ) -> None:
+        super().__init__()
+        self.tokenizer = tokenizer
+        self.pad_id = self.tokenizer.encode("[PAD]").ids[0]  # type: ignore
+        self.vocab_size = self.tokenizer.get_vocab_size()  # type: ignore
+        self.context_len = context_len
+        # MUGEN treats padding as unique ids so adding them to the total text tokens
+        # https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/gpt/gpt.py#L44
+        self.num_text_tokens = self.vocab_size + context_len
+        self.embedding = nn.Embedding(self.num_text_tokens, d_model)
+
+    def text_to_tokens(self, sentences: List[str]) -> Tensor:
+        """Pads the sentences to be of equal lengths"""
+        tokens = [
+            self.tokenizer.encode(sentence.strip().lower() + " [SEP]")  # type: ignore
+            for sentence in sentences
+        ]
+        token_ids = [t.ids[: self.context_len] for t in tokens]
+        # pad each sentence to be of length `context_len`
+        for i, t in enumerate(token_ids):
+            t += [self.pad_id] * (self.context_len - len(t))
+            token_ids[i] = t
+
+        return torch.Tensor(token_ids).type(torch.int64)
+
+    def encode(self, sentences: List[str], device: str) -> Tensor:
+        """Encodes sentences to token ids"""
+        token_ids = self.text_to_tokens(sentences).to(device)
+        # bump padding token ids by vocab_size so that they do not conincide with un-padded token ids
+        # and that the padding token ids themselves are unique
+        unique_pad_ids = torch.arange(self.context_len, device=device) + self.vocab_size
+        token_ids = torch.where(token_ids == self.pad_id, unique_pad_ids, token_ids)
+        return token_ids
+
+    def _filter_token_ids(self, token_ids: List[int]) -> List[Optional[int]]:
+        """Filters out token ids out side of vocab"""
+        return [
+            token_id
+            for token_id in token_ids
+            if token_id > 0 and token_id <= self.vocab_size
+        ]
+
+    def decode(self, token_ids: Tensor) -> List[str]:
+        """Decodes token ids back to sentences"""
+        sentences = []
+        for _token_ids in token_ids:  # iterate over batches
+            _token_ids = self._filter_token_ids(_token_ids.tolist())
+            sentence = self.tokenizer.decode(_token_ids)  # type: ignore
+            sentences.append(sentence)
+
+        return sentences
+
+    def lookup(self, token_ids: Tensor) -> Tensor:
```

Comment:
Just wondering, why do we include embeddings and lookup in this class? Seems without them we would basically just have a tokenizer class and and embedding module, then we could bundle these together in MultimodalGPT and just use the embedding directly there. (Or is there some case where we need both the tokenizer and the embedding logic in the same class?)

Reply:
This is a design choice to align the APIs of different tokenizers to be suitable for generation.

`MultimodalGPT` requires the modality-specific tokenizers to implement the same [APIs](https://github.com/facebookresearch/multimodal/blob/258b5b573d5040d01d5eb54b754a44243c9afaf0/torchmultimodal/models/gpt.py#L235-L303):
- `encode`: data to token ids
- `lookup`: token ids to embedding vectors
- `decode`: token ids to data

Although a text tokenizer is strictly for tokenization, it needs to have the above APIs to be consistent with the video tokenizer (VQVAE) where tokenization and embedding (2 in 1) both come from `Codebook`:

https://github.com/facebookresearch/multimodal/blob/258b5b573d5040d01d5eb54b754a44243c9afaf0/torchmultimodal/modules/layers/codebook.py#L261-L263